src/board.py

```diff
@@ -140,3 +141,33 @@ def square_of(self, piece: Piece | None) -> Square | None:
             for file, board_square in enumerate(board_rank):
                 if board_square is piece:
                     return Square(Vector(rank, file))  # HACK: I do not like that I have to chain constructors like this.
+
+    
+    def _make_condition(self, piece: Piece, captured: bool) -> Callable[[Square], bool]:
+        """Create a condition function
+        """
+        def condition(target_square: Square) -> bool:
+            other_piece = None
+            if captured:
+                return False
+            try:
+                other_piece = self[target_square]
+            except IndexError:
+                return False
+            return other_piece is None or piece.color != other_piece.color
+        return condition
+        
+    
+    def list_moves(self, selected_square: Square) -> list[tuple[Square, Piece]]:
+        piece = self[selected_square]
+        captured = False
+        condition = self._make_condition(self[selected_square], captured)
+        moves = []
+        for target_square in piece.legal_moves(selected_square, condition):
+            other_piece = self[target_square]
+            if other_piece is not None and other_piece.color != piece.color:
+                captured = True
+                condition = self._make_condition(piece, captured)
+            moves.append(target_square)
```

Comment:
You probably mean `moves.append((target_square, other_piece))` here? In accordance to your output signature ` -> list[tuple[Square, Piece]]`. I strongly advise against using `list` and use 'set' instead. Even in this format, `tuple[Square, Piece]` objects are expected and shall be unique in a chess game per piece, or even per player.

Reply:
I guess there is a chance that the same move is recorded twice? I can change that, no worries.

Reply:
Yes, I recon a `list` has the risk of bloating with repetitions, not necessarily, but since semantically speaking this is indeed a set, so I guess why not model it as such. As a bonus, a set is hashable.

Reply:
A set is not hashable, a frozenset is. Also, now Pieces need to be Hashable as well.